### ISSUE TITLE
Failed to get objects: Unable to connect to the in backend...

### ISSUE THREAD

**Issue by ArielGM91** [NagVis/nagvis#246]
Hi.
I recently updated my version of ChecK_MK (OMD package) from version 1.5.0p2 to 1.6.0p8
I have configured distributed monitoring with a slave server where I enabled encryption using TLS on the connection and checking the “Verify the Livestatus server certificate using the local site CA” checkbox.
All good until this step, because Distributed Monitoring works well.

But in Nagvis when I want to get a host from that slave server, it shows me the following error seen in the screenshot:
![Screenshot_5](https://user-images.githubusercontent.com/62817734/77835779-7e4ea280-7126-11ea-997d-8f44b6d4b993.jpg)

Nagvis version: 1.9.18
Distribution: Red Hat Enterprise Linux Server 7.3
Apache version: 2.4.6
PHP version: 5.4.16
Browser used: Mozilla Firefox, Google Chrome.
OMD Master version: 1.6.0p8
OMD Slave version: 1.6.0p6

Please help me with this. I have tried several options but they have no effect.

**Comment by AnsgarKlein** [comment]
I believe I have the same problem.
Every connection to mklivestatus backend fails with an SSL error. Creating new objects as well as existing objects on existing map.

![nagvis_mklivestatus](https://user-images.githubusercontent.com/1733578/104742974-6bf03480-574b-11eb-8f3c-3e5176f87b58.png)

I'm runnning (on both Livestatus servers):
Checkmk 1.6.0p17
Distribution: Red Hat Enterprise Linux Server 7.9 compatible
Apache: 2.4.6
PHP: 5.4.16

Sadly I am not sure how to check what version of nagvis is included in my checkmk package. I couldn't find any version string in the source code.

`etc/nagvis/conf.d/cmk_backends.ini.php` includes the following for that backend:
```
[backend_kng]
backendtype="mklivestatus"
socket="tcp-tls:XXX.XXX.XXX.XXX:6560"
statushost="XXXXXXXXXXXXXXXXXX"
verify_tls_peer=1
verify_tls_ca_path=/omd/sites/XXXXXXX/var/ssl/ca-certificates.crt
```

Testing with openssl works when using TLS1.2. Using anything other throws a similar error:
```
openssl s_client -connect XXX.XXX.XXX.XXX:6560 -tls1_1
CONNECTED(00000003)
139900438591376:error:1409442E:SSL routines:ssl3_read_bytes:tlsv1 alert protocol version:s3_pkt.c:1493:SSL alert number 70
139900438591376:error:1409E0E5:SSL routines:ssl3_write_bytes:ssl handshake failure:s3_pkt.c:659:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 0 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.1
    Cipher    : 0000
    Session-ID: 
    Session-ID-ctx: 
    Master-Key: 
    Key-Arg   : None
    Krb5 Principal: None
    PSK identity: None
    PSK identity hint: None
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
---
```

Is there any way to configure Nagvis / MKLivestatus to connect to the backend using TLS1.2 ?
I believe there is no configuration option to make Checkmk Livestatus server available under anything other than TLS1.2.

**Comment by AnsgarKlein** [comment]
> 
> Sadly I am not sure how to check what version of nagvis is included in my checkmk package. I couldn't find any version string in the source code.
> 

I updated to Checkmk v1.6.0p20 (Raw Edition) (latest stable version) which [changelog](https://checkmk.com/de/werk/11494) mentions using Nagvis 1.9.23.

Unfortunately this did not fix the error

**Comment by AnsgarKlein** [comment]
I can now confirm the TLS version is the issue.

I'm runnning:
Checkmk 1.6.0p20 (includes Nagvis 1.9.23)
Distribution: Red Hat Enterprise Linux Server 7.9 compatible
Apache: 2.4.6
PHP: 5.4.16

My `etc/nagvis/conf.d/cmk_backends.ini.php` on the connecting server includes:
```
[backend_<name>]
backendtype="mklivestatus"
socket="tcp-tls:<ip>:<port>"
statushost="[...]"
verify_tls_peer=1
verify_tls_ca_path=[...]/ca-certificates.crt
```

The stunnel on the receiving server has the following (default Checkmk) configuration:
```
cert = [...]
pid = [...]

output = [...]
syslog = no
# Set log level to warning
debug = 4

sslVersion = TLSv1.2
socket = l:TCP_NODELAY=1
socket = r:TCP_NODELAY=1
client = no

[live-tls]
accept = [...]
connect = [...]
TIMEOUTclose = 0
```

Changing `sslVersion = TLSv1.2` as a workaround to:
```
options = NO_SSLv2
options = NO_SSLv3
options = NO_TLSv1.1
sslVersion = all
```
(which only allows TLS-1.0 and TLS-1.2)

allows Nagvis to connect successfully. Also adding `options = NO_TLSv1` makes it so that Nagvis cannot connect again.

## tl;dr

Nagvis uses TLS-1.0 for connecting to TLS-Livestatus using *mklivestatus* backend, which by default only accepts TLS-1.2.  
Please support TLS-1.2 in Nagvis.

It would also be nice if Lars could pass this on to his employer, since Checkmk includes Nagvis in a way that will break when enabling encrypted Livestatus.
(We are not a paying customer so as far as I know there is no way for me to contact them)

**Comment by LarsMichelsen** [comment]
To me it seems like an incompatibility caused by the PHP version that is available on your systems. The PHP 5.4 does not seem to support TLSv1.2 while the default settings of Livestatus TLS enforce TLSv1.2 to be used.

For me there seem to be 2 options to solve this locally:

a) Update PHP to more recent version that is able to speak TLSv1.2. You will need at least PHP 5.6 (which seems to be available via SCL for RedHat) and makes it work out of the box.
b) Change livestatus TLS settings to support older TLS versions as AnsgarKlein suggested.

On one hand I understand that you'd like to have a out-of-the-box working solution with the common enterprise distros. On the other hand I don't think that it would be a good idea to lower the TLS security just for supporting those old PHP clients. PHP 5.4 is pretty outdated and an update to at least 5.6 seems to be doable. So i'd suggest to go this way.

**Comment by AnsgarKlein** [comment]
Thank you for your help!  
I can confirm that upgrading PHP to 7.4 fixes the TLS incompatibility.

I agree that upgrading PHP is a better way to handle this than to decrease security. I just wasn't aware of a minimum PHP version requirement for Nagvis (or Checkmk).

I think you can close this issue as there are two options to work around this problem.
